### PR TITLE
Implemented signup Api endpoint

### LINKED ISSUE #2: UserController: API to sign up
1. signup - "/user/signup"

 This endpoint is used to register a new user in the Quora Application.

- It should be a POST request

- This endpoint requests for all the attributes in 'SignupUserRequest' about the user.

- If the username provided already exists in the current database, throw ‘SignUpRestrictedException’ with the message code -'SGR-001' and message - 'Try any other Username, this Username has already been taken'.

- If the email Id provided by the user already exists in the current database, throw ‘SignUpRestrictedException’ with the message code -'SGR-002' and message -'This user has already been registered, try with any other emailId'.

- If the information is provided by a non-existing user, then save the user information in the database and return the 'uuid' of the registered user and message 'USER SUCCESSFULLY REGISTERED' in the JSON response with the corresponding HTTP status. Also, make sure to save the password after encrypting it using 'PasswordCryptographyProvider' class given in the stub file.

- Also, when a user signs up using this endpoint then the role of the person will be 'nonadmin' by default. You can add users with 'admin' role only by executing database queries or with pgAdmin.
-Read more @https://learn.upgrad.com/v/course/867/session/140118/segment/759581

-

### PR DESCRIPTION
**Pull request by soniyakocher**

### CODE REVIEW

**Review by cushivas — APPROVED** [review]
Looks good

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.UUID;
+
```
Comment (Piyasi-lab) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java:17]: please add javadoc style comments about class

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+@RestController
+@RequestMapping("/")
+public class UserController {
+
+    @Autowired
+    private UserBusinessService userBusinessService;
+
+
+    //The userSignup()  method is of request post type , path is "/user/signup" the method set all the attribute of user in userEntity and generate a Response model with HTTP status indicating successfully registered
```
Comment (Piyasi-lab) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java]: please change this comment to javadoc style comment

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        userEntity.setFirstName(signupUserRequest.getFirstName());
+        userEntity.setLastName(signupUserRequest.getLastName());
+        userEntity.setUsername(signupUserRequest.getUserName());
+        userEntity.setEmail(signupUserRequest.getEmailAddress());
+        userEntity.setPassword(signupUserRequest.getPassword());
+        userEntity.setContactNumber(signupUserRequest.getContactNumber());
+        userEntity.setAboutMe(signupUserRequest.getAboutMe());
+        userEntity.setDob(signupUserRequest.getDob());
+        userEntity.setCountry(signupUserRequest.getCountry());
+        userEntity.setSalt("1234abc");
```
Comment (Piyasi-lab) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java:47]: Should we hardcode salt and role? If we have to then declare it as private static variable

File: quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.context.request.WebRequest;
+
+
+// This annotation allow to write a code which will be made available to all controller class
+@ControllerAdvice
+
```
Comment (Piyasi-lab) [inline quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java]: Please remove extra whitespace and add javadoc style comment for this class

File: quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+public class RestExceptionHandler {
+
+    // The @ExceptionHandler allow you to define a method to handle specific exception
+    @ExceptionHandler(SignUpRestrictedException.class)
+
```
Comment (Piyasi-lab) [inline quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java]: Please add javadoc style comment for method and remove extra whitespaces

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
+package com.upgrad.quora.service.Entity;
+
+import org.apache.commons.lang3.builder.*;
+
+import javax.persistence.*;
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+import java.io.Serializable;
+@Entity
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: Please leave one line between imports and @Entity annotation

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+@Table(name ="users")
+
+
+// Created Jpql query used to retrive the data from db
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: Please add javadoc style comment

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+@NamedQueries({
+        @NamedQuery(name = "userByEmail", query = "select u from UserEntity u where u.email = :email"),
+        @NamedQuery(name= "userByUsername",query = "select u from UserEntity u where u.username=:username")
+})
+
+
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: Please remove extra whitespaces

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    public String getLastName() {
+        return lastName;
+    }
+
+    public void setLastName(String lastName) {
+        this.lastName = lastName;
+    }
+
+
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: Please remove extra white spaces

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.upgrad.quora.service.Entity.UserEntity;
+import com.upgrad.quora.service.dao.UserDao;
+import com.upgrad.quora.service.exception.SignUpRestrictedException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:11]: Please add javadoc style comments about class

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+public class UserBusinessService {
+
+    @Autowired
+    private UserDao userDao;
+
+    //The signup method is used to register a user by calling the createUser() method of UserDao class
+    //This method also generate SignUPRestrictedException to check if the username or email already present in database
+    //It gives the error message if the username or email has already taken or used by some other user and if not it register the user
+    //This method also encrypt the password and set password and generate salt in encrypted format
+    @Transactional(propagation = Propagation.REQUIRED)
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:24]: Please change this comment to javadoc style

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+            throw new SignUpRestrictedException("SGR-001", "Try any other Username, this Username has already been taken");
+        }
+
+        UserEntity existingUserEntityEmail = userDao.getUserByEmail((userEntity.getEmail()));
+        if (existingUserEntityEmail != null) {
+            throw new SignUpRestrictedException("SGR-002", "This user has already been registered, try with any other emailId");
+        }
+
+        String password = userEntity.getPassword();
+        if (password == null) {
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:42]: we should also check if the password is empty or contains just whitespaces

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import com.upgrad.quora.service.Entity.UserEntity;
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.NoResultException;
+import javax.persistence.PersistenceContext;
+
+@Repository
+public class UserDao {
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java:12]: Please add javadoc style comments

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+    //@persistenceContext annotation is used to change the state of userEntity from transient to persist
+    @PersistenceContext
+    private EntityManager entityManager;
+
+
+    //The getUserByUsername() method is used to check if there  already exist a user with same email
+    public UserEntity getUserByEmail(final String email) {
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java:24]: Please add javadoc style comment

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        try {
+            return entityManager.createNamedQuery("userByEmail", UserEntity.class).setParameter("email", email).getSingleResult();
+        } catch (NoResultException nre) {
+            System.out.println(nre.getMessage());
+            return null;
+        }
+    }
+
+    //The getUserByUsername() method is used to check if there  already exist a user with same username
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]: Please add javadoc style comments

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        try {
+            return entityManager.createNamedQuery("userByUsername", UserEntity.class).setParameter("username", username).getSingleResult();
+        } catch (NoResultException nre) {
+            System.out.println(nre.getMessage());
+            return null;
+        }
+    }
+
+
+    //This method is used to register a user in data base by storing all the user information
```
Comment (Piyasi-lab) [inline quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]: Please add javadoc style comments

**Review by Piyasi-lab — CHANGES_REQUESTED** [review]

**Comment by Piyasi-lab** [comment]
Also please uncomment corresponding tests

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java (hunk 3 above)
Comment (soniyakocher) [reply quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java:47]: 1.  when password encryption is applied we are modifying  this variable as userEntity.setSalt(encryptedText[0]); so a method which is to be overridden or modify  cannot be declare static.
2. hard coding role is given in the instruction to set the role as "nonadmin" by default for all users

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java (hunk 1 above)
Comment (soniyakocher) [reply quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java:17]: Changes are done

File: quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java (hunk 2 above)
Comment (soniyakocher) [reply quora-api/src/main/java/com/upgrad/quora/api/controller/UserController.java]: Done the changes

File: quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java (hunk 4 above)
Comment (soniyakocher) [reply quora-api/src/main/java/com/upgrad/quora/api/exception/RestExceptionHandler.java]: done

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java (hunk 7 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: changes are done

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java (hunk 8 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: changed

File: quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java (hunk 9 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/Entity/UserEntity.java]: Done changes.

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java (hunk 10 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:11]: Change are done

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java (hunk 12 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:42]: working on this

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java (hunk 14 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java:24]: added the javadoc style comment

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java (hunk 15 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]: Added

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java (hunk 16 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java]: Added comments

File: quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java (hunk 13 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/dao/UserDao.java:12]: Added the javadoc style comment

File: quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java (hunk 11 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/business/UserBusinessService.java:24]: Added javadoc style comment

**Review by Piyasi-lab — APPROVED** [review]

**Review by swapdutt — APPROVED** [review]
Looks fine to me

**Pull request merged**